crates/oxc_linter/src/rules/react/only_export_components.rs

```diff
@@ -0,0 +1,930 @@
+    ///
+    /// Check `.js` files that contain JSX (in addition to `.tsx`/`.jsx`). To reduce
+    /// false positives, only files that import React are checked when this is enabled.
+    ///
+    /// ```json
+    /// {
+    ///   "react/only-export-components": ["error", { "checkJS": true }]
+    /// }
+    /// ```
+    OnlyExportComponents,
+    react,
+    restriction
+);
+
+static DEFAULT_REACT_HOCS: &[&str] = &["memo", "forwardRef"];
+
+impl Rule for OnlyExportComponents {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let config = value
+            .as_array()
+            .and_then(|arr| arr.first())
+            .and_then(|first| {
+                serde_json::from_value::<OnlyExportComponentsOptionsJson>(first.clone()).ok()
+            })
+            .map(|config| OnlyExportComponentsConfig {
+                allow_export_names: config
+                    .allow_export_names
+                    .unwrap_or_default()
+                    .into_iter()
+                    .collect(),
+                allow_constant_export: config.allow_constant_export.unwrap_or_default(),
+                custom_hocs: config.custom_hocs.unwrap_or_default(),
+                check_js: config.check_js.unwrap_or_default(),
+            })
+            .unwrap_or_default();
+
+        Self(Box::new(config))
+    }
+
+    fn should_run(&self, ctx: &crate::context::ContextHost) -> bool {
+        let Some(filename) = ctx.file_path().file_name().and_then(|s| s.to_str()) else {
+            return false;
+        };
+
+        Self::should_scan(filename, self.check_js) && !Self::should_skip(filename)
+    }
+
+    fn run_once(&self, ctx: &LintContext<'_>) {
+        let react_is_in_scope = ctx
+            .module_record()
+            .import_entries
+            .iter()
+            .any(|entry| entry.module_request.name == "react");
+        if self.check_js && !react_is_in_scope {
+            return;
+        }
+
+        let react_hocs: Vec<&str> = DEFAULT_REACT_HOCS
+            .iter()
+            .copied()
+            .chain(self.custom_hocs.iter().map(String::as_str))
+            .collect();
+
+        let export_info = self.analyze_exports(ctx, &react_hocs);
+        let local_components = Self::find_local_components(ctx, &react_hocs);
+
+        Self::report_diagnostics(ctx, &export_info, &local_components);
+    }
+}
+
+impl OnlyExportComponents {
+    fn should_skip(f: &str) -> bool {
+        f.contains(".test.")
+            || f.contains(".spec.")
+            || f.contains(".cy.")
+            || f.contains(".stories.")
+    }
+
+    fn should_scan(f: &str, check_js: bool) -> bool {
+        use std::path::Path;
+        Path::new(f).extension().is_some_and(|ext| ext.eq_ignore_ascii_case("tsx"))
+            || Path::new(f).extension().is_some_and(|ext| ext.eq_ignore_ascii_case("jsx"))
+            || (check_js
+                && Path::new(f).extension().is_some_and(|ext| ext.eq_ignore_ascii_case("js")))
+    }
```

Comment:
```suggestion
    fn should_scan(f: &str, check_js: bool) -> bool {
        use std::path::Path;
        let ext = Path::new(f).extension().and_then(|e| e.to_str());

        matches!(ext, Some(e) if e.eq_ignore_ascii_case("tsx") || e.eq_ignore_ascii_case("jsx"))
            || (check_js && matches!(ext, Some(e) if e.eq_ignore_ascii_case("js")))
    }
```

we can reduce the number of `Path:;new` calls by restructuring this slightly